spring-cloud-stream/src/main/java/org/springframework/cloud/stream/converter/CompositeMessageConverterFactory.java

```diff
@@ -65,56 +65,46 @@ public CompositeMessageConverterFactory(List<? extends AbstractFromMessageConver
-		this.converters.add(new JsonToPojoMessageConverter(objectMapper));
-		this.converters.add(new PojoToJsonMessageConverter(objectMapper));
-		this.converters.add(new ByteArrayToStringMessageConverter());
-		this.converters.add(new StringToByteArrayMessageConverter());
-		this.converters.add(new PojoToStringMessageConverter());
-		this.converters.add(new JavaToSerializedMessageConverter());
-		this.converters.add(new SerializedToJavaMessageConverter());
+		this.converters.add(new TupleJsonMessageConverter(this.objectMapper));
+
+		MappingJackson2MessageConverter jsonMessageConverter = new MappingJackson2MessageConverter();
+		jsonMessageConverter.setSerializedPayloadClass(String.class);
+		if (this.objectMapper != null) {
+			jsonMessageConverter.setObjectMapper(this.objectMapper);
+		}
+
+		this.converters.add(jsonMessageConverter);
+		this.converters.add(new ByteArrayMessageConverter());
+
+		this.converters.add(new StringMessageConverter());
+		this.converters.add(new JavaSerializationMessageConverter());
 	}
 
 	/**
 	 * Creation method.
-	 * @param targetMimeType the target MIME type
+	 * @param mimeType the target MIME type
 	 * @return a converter for the target MIME type
 	 */
-	public CompositeMessageConverter getMessageConverterForType(MimeType targetMimeType) {
-		List<MessageConverter> targetMimeTypeConverters = new ArrayList<MessageConverter>();
-		for (AbstractFromMessageConverter converter : converters) {
-			if (converter.supportsTargetMimeType(targetMimeType)) {
-				targetMimeTypeConverters.add(converter);
+	public CompositeMessageConverter getMessageConverterForType(MimeType mimeType) {
+		List<MessageConverter> converters = new ArrayList<>();
+		for (MessageConverter converter : this.converters) {
+			if (converter instanceof AbstractMessageConverter) {
+				for (MimeType type : ((AbstractMessageConverter) converter).getSupportedMimeTypes()) {
+					if (type.includes(mimeType)) {
+						converters.add(converter);
+					}
+				}
 			}
 		}
-		if (CollectionUtils.isEmpty(targetMimeTypeConverters)) {
+		if (CollectionUtils.isEmpty(converters)) {
 			throw new ConversionException("No message converter is registered for "
-					+ targetMimeType.toString());
+					+ mimeType.toString());
 		}
-		return new CompositeMessageConverter(targetMimeTypeConverters);
+		return new CompositeMessageConverter(converters);
 	}
 
 	public CompositeMessageConverter getMessageConverterForAllRegistered() {
-		return new CompositeMessageConverter(new ArrayList<MessageConverter>(converters));
+		return new CompositeMessageConverter(new ArrayList<>(this.converters));
 	}
 
-	public Class<?>[] supportedDataTypes(MimeType targetMimeType) {
-		Set<Class<?>> supportedDataTypes = new HashSet<>();
-		// Make sure to check if the target type is of explicit java object type.
-		if (MessageConverterUtils.X_JAVA_OBJECT.includes(targetMimeType)) {
-			supportedDataTypes.add(MessageConverterUtils.getJavaTypeForJavaObjectContentType(targetMimeType));
```

Comment:
We can restore that behaviour but I wonder if it wouldn't be better to use `contentType` as a real `MimeType` and perhaps have a separate alternate parameter for `javaType`.